src/datanode/src/heartbeat/handler/open_region.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use catalog::{CatalogManagerRef, RegisterTableRequest};
+use common_catalog::format_full_table_name;
+use common_meta::instruction::{Instruction, InstructionReply};
+use common_telemetry::{error, warn};
+use snafu::ResultExt;
+use table::engine::manager::TableEngineManagerRef;
+use table::engine::EngineContext;
+use table::requests::OpenTableRequest;
+
+use crate::error::{self, Result};
+use crate::heartbeat::handler::HeartbeatResponseHandler;
+use crate::heartbeat::HeartbeatResponseHandlerContext;
+
+#[derive(Clone)]
+pub struct OpenTableHandler {
+    catalog_manager: CatalogManagerRef,
+    table_engine_manager: TableEngineManagerRef,
+}
+
+impl HeartbeatResponseHandler for OpenTableHandler {
+    fn is_acceptable(&self, ctx: &HeartbeatResponseHandlerContext) -> bool {
+        matches!(
+            ctx.incoming_message.as_ref(),
+            Some((_, Instruction::OpenRegion { .. }))
+        )
+    }
+
+    fn handle(&self, ctx: &mut HeartbeatResponseHandlerContext) -> Result<()> {
+        match ctx.incoming_message.take() {
+            Some((
+                meta,
+                Instruction::OpenRegion {
+                    catalog,
+                    schema,
+                    table,
+                    table_id,
+                    engine,
+                    region_number,
+                },
+            )) => {
+                ctx.finish();
+
+                let mailbox = ctx.mailbox.clone();
+                let self_ref = Arc::new(self.clone());
+
+                common_runtime::spawn_bg(async move {
```

Comment:
suggestion: wrap the code inside the spawn_bg into a function to make the `handle()` method body tidy.